### ISSUE TITLE
Constant redefinition detected in mutually exclusive blocks

### ISSUE THREAD

**Issue by ItsDrike** [microsoft/pyright#7738]
**Describe the bug**
Pyright has an optional `reportConstantRedefinition` option, which when enabled, makes pyright report any constant redefinitions, i.e. redefinitions of SCREAMING_SNAKE_CASE variables.

However, it seems that pyright ignore's the program's control flow while checking for this, and reports a redefinition even if it occurs in mutually exclusive blocks, like if-else blocks. This seems incorrect, as it isn't a redefinition at that point, it's just a conditional definition of the constant.

**Code or Screenshots**

```python
DEBUG = get_config("DEBUG", cast=bool, default=False)

if DEBUG:
  SECRET_KEY = "django-insecure-..."
else:
  SECRET_KEY = get_config("SECRET_KEY")  # Pyright: "SECRET_KEY" is constant (because it is uppercase) and cannot be redefined [reportConstantRedefinition]
```

**Comment by erictraut** [comment]
This behavior is correct. A constant, by definition, can be assigned a value only once and only on one branch. It's the equivalent of a `Final` variable.

If you don't want this rule to be enforced, you can disable the `reportConstantRedefinition` check.

**Comment by ItsDrike** [comment]
Huh, that is surprising, I do understand that a constant can only be assigned once, but the rule of it only happening in one branch is rather odd, and I don't really get the reason behind this rule. Especially when ternary conditions work just fine. Could you maybe expand a bit on why this should be problematic?

I quickly glanced over the type spec that talks about Final (<https://typing.readthedocs.io/en/latest/spec/qualifiers.html#uppercase-final>), and I didn't see it mention anything about the one branch rule (though I might have just missed that). In my opinion, code like this should be valid:

```python
test: Final[int]

if DEBUG:
    test = 1
else:
    test = 2
```

Is there something I missed in the PEP? Or is there some reason why this code shouldn't be valid? The assignment is still guaranteed to only happen once here.
